aesara/tensor/nlinalg.py

```diff
@@ -823,6 +824,69 @@ def tensorsolve(a, b, axes=None):
     return TensorSolve(axes)(a, b)
 
 
+class MatMulOp(Op):
+    __props__ = ("dtype",)
+
+    def __init__(self, dtype=None):
+        self.dtype = dtype
+
+    def make_node(self, a, b):
+        a = at.as_tensor_variable(a)
+        b = at.as_tensor_variable(b)
+
+        if 0 in (a.ndim, b.ndim):
+            raise ValueError("inputs to `matmul` cannot be scalar.")
+
+        if 2 == a.ndim == b.ndim:
+            out = matrix(dtype=self.dtype)
```

Comment:
`broadcastable` indicates which dimensions have shape `1`.  This light form of shape inference can be pretty important for optimizations, so it's always worth attempting to determine the `broadcastable` pattern given the `broadcastable` patterns of the inputs.

Take a look at `aesara.tensor.math.Dot.make_node` for some of the relevant logic.